### PR TITLE
#141404671 add Player communities

### PR DESCRIPTION
**Pull request by jomadoye**
#### What does this PR do?
This PR adds the players community feature to the app

#### Description of Task to be completed?
Registered users should be able to invite other users to connect as "friends"

#### How should this be manually tested?
Click the `invite friends` button on the game page

#### Any background context you want to provide?
- Users can add friends.
- Users can delete friends.

#### What are the relevant pivotal tracker stories?
#141404671

#### Screenshots (if appropriate)
screen shot 2017-04-07 at 12 58 50 pm
screen shot 2017-04-07 at 12 59 01 pm
screen shot 2017-04-07 at 12 59 11 pm

#### Questions:
N/A

### CODE REVIEW

**Comment by tonymontaro** [comment]
Great job Jed! Interact with the app here (review server): http://kakashi-cfh-review.herokuapp.com/

**Review by tezebuike — COMMENTED** [review]
Nice work... LGTM

**Comment by fmustapha** [comment]
Looks good!

File: public/views/question.html [hunk 1]
```diff
@@ -1,6 +1,8 @@
 <div id="question-container-inner" class="animate slideInRight">
   <div class="card longBlack">
     <span ng-bind-html="game.curQuestion.text" ng-hide="game.state === 'game ended' || game.state ==='game dissolved' || game.state === 'awaiting players'"></span>
+    <button class="btn btn-donate" id="inviteFriendsBtn" ng-click="inviteFriends()">Invite Friends</button>
```
Comment (tonymontaro) [inline public/views/question.html]: We should hide the 'Invite Friends' button if the users is not logged in.

File: app/controllers/users.js [hunk 2]
```diff
@@ -44,6 +44,99 @@ exports.findUser = (req, res) => {
   });
 };
 
+exports.getFriends = (req, res) => {
+  const userId = req.body.userId;
+  try {
```
Comment (sdowoade) [inline app/controllers/users.js]: Why the use of a try catch block around functions with callbacks? We should return responses with appropriate status codes and message when there is a success or error.

File: public/views/question.html (hunk 1 above)
Comment (Andela-JUdensi) [inline public/views/question.html]: That has been handled. But will be available on the next PR

Comment (jomadoye) [reply public/views/question.html]: Thanks for pointing this out, i have fixed this issue.

File: app/controllers/users.js (hunk 2 above)
Comment (jomadoye) [reply app/controllers/users.js]: Thanks for pointing this out, i have replaced all the try, catch blocks with the appropraite response messages

File: app/controllers/users.js [hunk 3]
```diff
@@ -44,6 +44,98 @@ exports.findUser = (req, res) => {
   });
 };
 
+exports.getFriends = (req, res) => {
+  const userId = req.body.userId;
```
Comment (sdowoade) [inline app/controllers/users.js]: Getting there. Can we add the status codes as well?

File: app/controllers/users.js [hunk 4]
```diff
@@ -44,6 +44,98 @@ exports.findUser = (req, res) => {
...
+    safe: true,
+    upsert: true
+  }, (error) => {
+    if (!(error)) {
+      return error;
+    }
+  });
+
+  User.findById(userId, (err, user) => {
+    if (err) throw err;
```
Comment (sdowoade) [inline app/controllers/users.js]: Probably should respond with appropriate status codes here as well instead of throwing an error.

**Comment by jomadoye** [comment]
@andela-dowoade Please kindly review.

File: app/controllers/users.js [hunk 5]
```diff
@@ -29,17 +29,132 @@ exports.authCallback = (req, res) => {
...
 exports.findUser = (req, res) => {
   const userid = req.params.userid;
   User.findById(userid, (err, oneUser) => {
     if (!err) {
-      res.send(oneUser);
+      res.status(200)
+        .json(oneUser);
     } else {
-      res.send('An error occurred');
+      res.status(401)
```
Comment (sdowoade) [inline app/controllers/users.js]: `401` means unauthorised which I am not sure is ideal in this case. You'll probably want to use a `400` or `500` in most cases for errors depending on what the case may be and a `404` if a particular resource is not found.

Tip: mongoose `findById` does not throw an error if the resource returned is empty. i.e `oneUser` in this case may still be null if `err` is null.

https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html

File: app/controllers/users.js [hunk 6]
```diff
@@ -29,17 +29,132 @@ exports.authCallback = (req, res) => {
...
+};
+
+exports.getFriends = (req, res) => {
+  const userId = req.body.userId;
+  User.findById(userId, (err, user) => {
+    if (!err) {
+      res.status(200)
+        .json(user.friends);
+    } else {
+      res.status(404)
```
Comment (sdowoade) [inline app/controllers/users.js]: Should probably be a `500`. Err could be anything ranging for a failure to connect to the db.

File: app/controllers/users.js [hunk 7]
```diff
@@ -29,17 +29,132 @@ exports.authCallback = (req, res) => {
...
+    }
+  });
+};
+
+exports.getFriends = (req, res) => {
+  const userId = req.body.userId;
+  User.findById(userId, (err, user) => {
+    if (!err) {
+      res.status(200)
+        .json(user.friends);
```
Comment (sdowoade) [inline app/controllers/users.js]: Probably want to check that user is defined, if not return a `404`. Mongoose `findById` wouldn't throw an error if the user does not exist.

File: app/controllers/users.js (hunk 6 above)
Comment (jomadoye) [reply app/controllers/users.js]: 👍🏽

File: app/controllers/users.js (hunk 7 above)
Comment (jomadoye) [reply app/controllers/users.js]: 👍🏽

**Review by sdowoade — APPROVED** [review]

**Pull request merged**